[FLINK-16996][table] Refactor planner and connectors to use new data structures

wuchong opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Travis CI to do that following [this guide](https://flink.apache.org/contributing/contribute-code.html#open-a-pull-request).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

Refactors existing code to use the new data structures interfaces.

## Brief change log

The commints in the order:
- [table-common] Add necessary methods to internal data structures
- [table-common] Add binary implementations of internal data structures
- [table-runtime-blink] Implement all the data structures and serializers around RowData
- [table-runtime-blink] Remove legacy data formats (BaseRow)
- [table-blink] Refactor planner and runtime to use new data structures
- [python] Refactor pyflink to use new data structures
- [parquet] Refactor parquet connector to use new data structures
- [orc] Refactor ORC connector to use new data structures
- [hive] Refactor Hive connector to use new data structures

Some notable changes: 
- In code generation, we hard cast `StringData` to `BinaryStringData`. This makes the code generator easily to generate opeartions based on string. The same to `RawValueData`. 
- Most methods of `Decimal` have been moved to `DecimalDateUtil`. So I also updated the code generation logic.
- Remove `RecordEqualiser#equalsWithoutHeader` interface. This method is used less and can be replaced by `RecordEqualiser#equals`. This can also avoid add `equalsWithoutHeader` method to the public API `GenericRowData`.


## Verifying this change

This change is covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: yes
  - The runtime per-record code paths (performance sensitive): yes
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

wuchong commented:
Hi @dianfu , could you help to review the python part? `RowData` and `ArrayDat` don't implement `TypeGetterSetter`, so I have to refactor the `ArrowFieldWriter` implementations.

KurtYoung commented on flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/data/binary/BinaryStringDataUtil.java:
```diff
@@ -134,49 +136,49 @@
 		str.ensureMaterialized();
 
-		if (precision > Decimal.MAX_LONG_DIGITS || str.getSizeInBytes() > Decimal.MAX_LONG_DIGITS) {
+		if (DecimalDataUtils.isByteArrayDecimal(precision) || DecimalDataUtils.isByteArrayDecimal(str.getSizeInBytes())) {
```
Is this right? why we check both `precision` and `sizeInBytes` with the same method?

wuchong replied:
cc @JingsongLi , do you know why we check both of them?

JingsongLi replied:
The case is, precision is 10 (less than `MAX_LONG_DIGITS`), but the string data is more than 10 precision maybe, we can use big decimal to convert, and then remove the extra decimals according to the precision.

KurtYoung commented on flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/data/util/MapDataUtil.java:
```diff
@@ -16,44 +16,34 @@
+	 * Converts a {@link MapData} into Java {@link Map}, the keys and values of the Java map
+	 * still holds objects of internal data structures.
+	 */
+	public static Map<Object, Object> convertToJavaMap(
```
only used by tests

wuchong replied:
This is also used in code generation: 
https://github.com/wuchong/flink/blob/744983929e464758baad98bf73cba9601eac48c5/flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/ScalarOperatorGens.scala#L2179

dianfu commented:
@wuchong The python part LGTM.

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/data/GenericRowData.java:
```diff
@@ -186,6 +187,20 @@ public RowData getRow(int pos, int numFields) {
+			if (i != 0) {
+				sb.append(",");
+			}
+			sb.append(StringUtils.arrayAwareToString(fields[i]));
```
Use `org.apache.flink.table.utils.EncodingUtils#objectToString`

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/AsyncTableFunction.java:
```diff
@@ -55,7 +55,7 @@
  *
  * <p>By default the result type of an evaluation method is determined by Flink's type extraction
- * facilities. Currently, only support {@link org.apache.flink.types.Row} and {@code BaseRow} as
+ * facilities. Currently, only support {@link org.apache.flink.types.Row} and {@code RowData} as
```
use: `{@link RowData}`

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/utils/LogicalTypeUtils.java:
```diff
@@ -36,6 +43,54 @@ public static LogicalType removeTimeAttributes(LogicalType logicalType) {
+			case RAW:
+				return RawValueData.class;
+			default:
+				throw new UnsupportedOperationException("Not support type: " + type);
```
nit: `Unsupported type:`

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/utils/LogicalTypeUtils.java:
```diff
@@ -36,6 +43,54 @@ public static LogicalType removeTimeAttributes(LogicalType logicalType) {
 	}
 
+	/**
+	 * Get internal(sql engine execution data formats) conversion class for {@link LogicalType}.
```
nit: `Returns the conversion class for the given {@link LogicalType} that is used by the table runtime.`

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/utils/LogicalTypeUtils.java:
```diff
@@ -36,6 +43,54 @@ public static LogicalType removeTimeAttributes(LogicalType logicalType) {
+	/**
+	 * Get internal(sql engine execution data formats) conversion class for {@link LogicalType}.
+	 */
+	public static Class<?> internalConversionClass(LogicalType type) {
```
nit: `toInternalConversionClass`

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/data/binary/BinaryFormat.java:
```diff
@@ -0,0 +1,71 @@
+/**
+ * Binary format spanning {@link MemorySegment}s.
+ */
+public interface BinaryFormat {
```
Missing `@Internal`

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/data/binary/LazyBinaryFormat.java:
```diff
@@ -0,0 +1,139 @@
+ * <p>It can lazy the conversions as much as possible. It will be converted into required form
+ * only when it is needed.
+ */
+public abstract class LazyBinaryFormat<T> implements BinaryFormat {
```
Missing `@Internal`

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/BinarySegmentUtils.java:
```diff
@@ -0,0 +1,1201 @@
+ * Utilities for binary data segments which heavily uses {@link MemorySegment}.
+ */
+@Internal
+public class BinarySegmentUtils {
```
Put this under `org.apache.flink.table.data.binary`. Make the class `final` with a private default constructor.

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/MurmurHashUtils.java:
```diff
@@ -0,0 +1,171 @@
+ * Murmur Hash. This is inspired by Guava's Murmur3_32HashFunction.
+ */
+@Internal
+public final class MurmurHashUtils {
```
put under `org.apache.flink.table.data.binary` add private default constructor

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/StringUtf8Utils.java:
```diff
@@ -0,0 +1,302 @@
+ * Utilities for String UTF-8.
+ */
+@Internal
+public class StringUtf8Utils {
```
put under `org.apache.flink.table.data.binary` make final with private default constructor

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/data/TypedSetters.java:
```diff
@@ -0,0 +1,90 @@
+ * used on the binary format such as {@link BinaryRowData}.
+ */
+@Internal
+public interface TypedSetters {
```
If this is mainly used for binary formats, put it in the binary package.

twalthr commented:
Thanks @wuchong for this massive PR. I took a look at the classes in `table-common`. Apart from my comments, they look good to me.

JingsongLi commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/CodeGenUtils.scala:
```diff
@@ -132,6 +136,15 @@ object CodeGenUtils {
    * Returns a term for representing the given class in Java code.
    */
   def typeTerm(clazz: Class[_]): String = {
+    if (clazz == classOf[StringData]) {
```
Could this put CodeGen at risk of missing cast?
If possible, I prefer cast in accessing its methods.

JingsongLi replied:
Consider `RowData` and others, using `StringData` more reasonable to me.

wuchong replied:
There are too many calls on the `BinaryStringData` in code generation now, if we use StringData in code generation, we have to refactor a lot of codes. And we don't benefit much from this effort. I created FLINK-17437 to track this, we can refactor this in the future.

JingsongLi commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/CodeGeneratorContext.scala:
```diff
@@ -678,7 +678,7 @@ class CodeGeneratorContext(val tableConfig: TableConfig) {
       case None =>
         val term = newName("typeSerializer")
-        val ser = InternalSerializers.create(t, new ExecutionConfig)
+        val ser = InternalSerializers.createInternalSerializer(t, new ExecutionConfig)
```
`createInternalSerializer` a little redundant

wuchong commented:
Thanks @KurtYoung @dianfu @twalthr for the quickly reviewing. I have addressed all the comments. 

Hi Timo, currently, it's hard to make all classes under `binary.` to be package-visible, because they will be used in serializers and code generation. But I think `MurmurHashUtils` and `StringUtf8Utils` can be, so I updated them.

JingsongLi commented:
There are still 100+ BaseRow in codes (variable name or something else), you can modify them all.

wuchong commented:
Thanks for the reviewing @JingsongLi . I have addressed the comments and renamed lagecy field names and method using BaseRow, SqlTimestamp and BinaryGeneric.

wuchong commented:
Thanks all for the reviewing. I have rebased the commits. Will merge this once builds passed.

JingsongLi commented:
Hi @wuchong , do you want to squash or not?  Here is my question:
Does it need to be able to compile for each commit to the master?

wuchong commented:
Hi @JingsongLi , I want to keep the splitted commits. From my point of view, squashing such a large PR into one commit is not good.

JingsongLi commented:
> Hi @JingsongLi , I want to keep the splitted commits. From my point of view, squashing such a large PR into one commit is not good.

So the answer is we can push commit with broken compilation.

wuchong commented:
@JingsongLi , as far as I know, the community doesn't have a rule to make sure independent commits can pass build, but there is a rule to separate commits if it is a big commit:

https://flink.apache.org/contributing/code-style-and-quality-pull-requests.html#separate-refactoring-cleanup-and-independent-changes

wuchong commented:
Travis is passed: https://travis-ci.org/github/wuchong/flink/builds/680843224

JingsongLi commented:
@wuchong go you.
